### PR TITLE
[Sparse Index] Integrate with `git status`

### PR DESCRIPTION
**Pull request by derrickstolee**
Replaces #367, which used the upstream commits directly. This instead includes a merge commit onto `vfs-2.32.0` so we can move forward even if the upstream version is modified.

This integrates the sparse-index with `git status`. There is a LOT of code here, and much of it makes future changes (#361 and #364) much simpler than they would be otherwise.

This merges into `features/sparse-index` which is intended to be released as an experimental release. The goal of reviewing this PR is to find "good enough for a limited release to dogfooders" and not the same scrutiny of upstream.

In particular, I am currently looking for an issue with directory/file conflicts at the sparse-checkout boundary. They don't reveal themselves until `git checkout`, and only in very rare cases that won't affect our dogfooders.

### CODE REVIEW

File: t/t1092-sparse-checkout-compatibility.sh [hunk 1]
```diff
@@ -262,13 +281,29 @@ test_expect_success 'diff --staged' '
...
 '
 
-test_expect_success 'diff with renames' '
+test_expect_success 'diff with renames and conflicts' '
 	init_repos &&
 
 	for branch in rename-out-to-out rename-out-to-in rename-in-to-out
 	do
 		test_all_match git checkout rename-base &&
 		test_all_match git checkout $branch -- .&&
```
Comment (jeffhostetler) [inline t/t1092-sparse-checkout-compatibility.sh]: typo:  dot space amp amp ??

File: unpack-trees.c [hunk 2]
```diff
@@ -1066,6 +1068,13 @@ static struct cache_entry *create_ce_entry(const struct traverse_info *info,
 	/* len+1 because the cache_entry allocates space for NUL */
```
Comment (jeffhostetler) [inline unpack-trees.c:1088]: I feel like there is a bug here, but I can't quite put my finger on it.   We added `alloc_len` above and `make_empty...` created the cache_entry with alloc_len+1 space.  Passing 'len+1' to `make_traverse_path` feels odd, but i think is right -- because the actual path fragment doesn't have the trailing slash.

Maybe it is just the `/* len+1 because... */` comment that is throwing me off.

File: unpack-trees.c [hunk 3]
```diff
@@ -1074,20 +1083,28 @@ static struct cache_entry *create_ce_entry(const struct traverse_info *info,
...
-				 unsigned long dirmask,
-				 struct cache_entry **src,
-				 const struct name_entry *names,
-				 const struct traverse_info *info)
+static int unpack_single_entry(int n, unsigned long mask,
+			       unsigned long dirmask,
+			       struct cache_entry **src,
+			       const struct name_entry *names,
+			       const struct traverse_info *info,
+			       int sparse_directory)
```
Comment (jeffhostetler) [inline unpack-trees.c:1111]: nit: maybe do this function rename in a separate commit??

File: unpack-trees.c [hunk 4]
```diff
@@ -1074,20 +1083,28 @@ static struct cache_entry *create_ce_entry(const struct traverse_info *info,
...
 		return 0;
 
+	/* defer work if our cache entry doesn't match the expectations. */
+	if (sparse_directory) {
+		if (src[0] && !S_ISSPARSEDIR(src[0]->ce_mode))
+			BUG("expected sparse directory entry");
+	} else if (src[0] && S_ISSPARSEDIR(src[0]->ce_mode)) {
+		return 0;
+	}
+
```
Comment (jeffhostetler) [inline unpack-trees.c:1127]: I think I understand what is happening here, but I'm still a little fuzzy.
Would the above read better if it was something like:

```
if (src[0]) {
    if (sparse_directory && !S_ISSPARSEDIR())
        BUG("...");
    if (!sparse_directory && S_ISSPARSEDDIR())
        return 0; /* defer work... */
}
```

I am curious about the !src[0] cases.  These represent conflicts.

File: unpack-trees.c [hunk 5]
```diff
@@ -1221,16 +1240,59 @@ static int find_cache_pos(struct traverse_info *info,
...
+ * Compare ce->name to p->path + '/' otherwise. Note that
+ * ce->name must end in a trailing '/' because it is a sparse
+ * directory entry.
+ */
+static int sparse_dir_matches_path(const struct cache_entry *ce,
+				   struct traverse_info *info,
+				   const struct name_entry *p)
+{
+	assert(S_ISSPARSEDIR(ce->ce_mode));
+	assert(ce->name[ce->ce_namelen - 1] == '/');
```
Comment (jeffhostetler) [inline unpack-trees.c:1275]: I like assert(), but they disappear in production builds.  Should we make these BUG()s.

File: unpack-trees.c [hunk 6]
```diff
@@ -1221,16 +1240,59 @@ static int find_cache_pos(struct traverse_info *info,
...
+
+	if (info->namelen)
+		return ce->ce_namelen == info->namelen + p->pathlen + 2 &&
+		       ce->name[info->namelen] == '/' &&
+		       !strncmp(ce->name, info->name, info->namelen) &&
```
Comment (jeffhostetler) [inline unpack-trees.c:1280]: do we care about case folding in these 3 strncmp()s?

File: unpack-trees.c [hunk 7]
```diff
@@ -1221,16 +1240,59 @@ static int find_cache_pos(struct traverse_info *info,
...
 	if (0 <= pos)
 		return o->src_index->cache[pos];
-	else
+
+	/*
+	 * Check for a sparse-directory entry named "path/".
+	 * Due to the input p->path not having a trailing
+	 * slash, the negative 'pos' value overshoots the
+	 * expected position by one, hence "-2" here.
+	 */
```
Comment (jeffhostetler) [inline unpack-trees.c:1301]: This makes me nervous.  (I've never liked the negative number trick used by these `find` functions.)

Could we add a test that includes "path" and "path-" and "path/" and "path0"?

I'm curious about "path-" since the dash is 0x2d and is before the slash at 0x2f.  And is a valid character in a Windows file name.  Most of the rest 0x00 thru 0x2e are not valid or not valid trailing characters.

Maybe also "path-foo".

File: dir.c [hunk 8]
```diff
@@ -1397,6 +1397,17 @@ enum pattern_match_result path_matches_pattern_list(
...
+	 * Directory entries are matched if and only if a file
+	 * contained immediately within them is matched. For the
+	 * case of a directory entry, modify the path to create
+	 * a fake filename within this directory, allowing us to
+	 * use the file-base matching logic in an equivalent way.
+	 */
+	if (parent_pathname.len > 0 &&
+	    parent_pathname.buf[parent_pathname.len - 1] == '/')
+		strbuf_add(&parent_pathname, "-", 1);
+
```
Comment (jeffhostetler) [inline dir.c:1469]: (_The PR view won't let me comment on something outside of a change block._)

On line 1417 we compute `slash_pos`.  If the code at 1409 added the "-", we also already know where the last slash is and could avoid the strrchr() below.

File: diff-lib.c [hunk 9]
```diff
@@ -325,6 +325,11 @@ static void show_new_file(struct rev_info *revs,
 	unsigned dirty_submodule = 0;
 	struct index_state *istate = revs->diffopt.repo->index;
```
Comment (jeffhostetler) [inline diff-lib.c:327]: I don't know enough about the diff code to comment on this.

I do wonder if there is enough info at line 360 to do a quick oidcmp() before actually opening them.

File: t/t1092-sparse-checkout-compatibility.sh [hunk 10]
```diff
@@ -508,12 +508,17 @@ test_expect_success 'sparse-index is expanded and converted back' '
...
+test_expect_success 'sparse-index is not expanded' '
+	init_repos &&
+
+	rm -f trace2.txt &&
+	echo >>sparse-index/untracked.txt &&
 	GIT_TRACE2_EVENT="$(pwd)/trace2.txt" GIT_TRACE2_EVENT_NESTING=10 \
-		git -C sparse-index -c core.fsmonitor="" status -uno &&
-	test_region index ensure_full_index trace2.txt
+		git -C sparse-index status &&
+	test_region ! index ensure_full_index trace2.txt
```
Comment (jeffhostetler) [inline t/t1092-sparse-checkout-compatibility.sh:523]: Nice test function!

File: wt-status.c [hunk 11]
```diff
@@ -657,6 +657,34 @@ static void wt_status_collect_changes_index(struct wt_status *s)
...
+{
+	struct string_list_item *it;
+	struct wt_status_change_data *d;
+	struct wt_status *s = context;
+	char *full_name;
+
+	if (S_ISDIR(mode))
+		return READ_TREE_RECURSIVE;
+
+	full_name = xstrfmt("%s%s", base->buf, path);
```
Comment (jeffhostetler) [inline wt-status.c]: nit: some people complain about the efficiency of xstrfmt in cases like this where we just have "%s%s".  We could just build it in a strbuf before inserting it.

Also, I think we're leaking full_name -- assuming that the insert does a copy.

File: sparse-index.c [hunk 12]
```diff
@@ -282,6 +286,9 @@ void ensure_full_index(struct index_state *istate)
 	istate->cache = full->cache;
 	istate->cache_nr = full->cache_nr;
 	istate->cache_alloc = full->cache_alloc;
+	istate->fsmonitor_has_run_once = 0;
+	FREE_AND_NULL(istate->fsmonitor_dirty);
+	FREE_AND_NULL(istate->fsmonitor_last_update);
```
Comment (jeffhostetler) [inline sparse-index.c:291]: nit: is there a typo on line 264 in the comment ?
should "cd->name" be "ce->name" ?

File: t/t7519-status-fsmonitor.sh [hunk 13]
```diff
@@ -383,4 +384,51 @@ test_expect_success 'status succeeds after staging/unstaging' '
...
+# check_sparse_index_behavior [!]
+# If "!" is supplied, then we verify that we do not call ensure_full_index
+# during a call to 'git status'. Otherwise, we verify that we _do_ call it.
+check_sparse_index_behavior () {
+	git status --porcelain=v2 >expect &&
+	git sparse-checkout init --cone --sparse-index &&
+	git sparse-checkout set dir1 dir2 &&
+	GIT_TRACE2_EVENT="$(pwd)/trace2.txt" GIT_TRACE2_EVENT_NESTING=10 \
+		git status --porcelain=v2 >actual &&
+	test_region $1 index ensure_full_index trace2.txt &&
```
Comment (jeffhostetler) [inline t/t7519-status-fsmonitor.sh:397]: The paranoid in me would add

```
test_region fsm_hook query trace2.txt &&
```

to confirm that the hook was invoked.  (And maybe verify that got a successful, non-trivial response.)
Mainly to verify that we received and are using data from the hook and did not fallback
to the non-fsmonitor code path.

**Review by jeffhostetler — APPROVED** [review]
I marked up a few things, but didn't find anything serious.  (I'm still a little fuzzy on some of the details, but that is to be expected.)

File: t/t1092-sparse-checkout-compatibility.sh (hunk 1 above)
Comment (derrickstolee) [reply t/t1092-sparse-checkout-compatibility.sh]: Looks like that bit is upstream already. I'll take a look into a cleanup sometime.

File: unpack-trees.c (hunk 2 above)
Comment (derrickstolee) [reply unpack-trees.c:1088]: I should update the comment.

File: unpack-trees.c (hunk 5 above)
Comment (derrickstolee) [reply unpack-trees.c:1275]: This is a thing Elijah has started doing in his merge-ORT stuff, so I've been using them for similar reasons. It communicates expectations in the context of the patch but also gives us some protection when running with debug bits.

File: unpack-trees.c (hunk 4 above)
Comment (derrickstolee) [reply unpack-trees.c:1127]: This is also the critical code that fails the d/f conflict case. I'll ping you later to gather some of your knowledge here.

File: unpack-trees.c (hunk 6 above)
Comment (derrickstolee) [reply unpack-trees.c:1280]: I believe that we don't at this point since we are interacting directly between an index and a tree. We are not interacting with user-specified pathspecs or the filesystem.

File: unpack-trees.c (hunk 2 above)
Comment (derrickstolee) [reply unpack-trees.c:1088]: oh! _and_ the method call underneath. Thanks!

File: unpack-trees.c (hunk 3 above)
Comment (derrickstolee) [reply unpack-trees.c:1111]: I'll think about it. It's definitely part of the d/f conflict situation, so I will handle that when I have the best patch for upstream.

File: unpack-trees.c (hunk 7 above)
Comment (derrickstolee) [reply unpack-trees.c:1301]: That's a good point. We might need to backtrack a bit to be sure. Will investigate.

Comment (derrickstolee) [reply unpack-trees.c:1301]: Yes, this matters. I can repro it in a test now.

File: diff-lib.c (hunk 9 above)
Comment (derrickstolee) [reply diff-lib.c:327]: We wouldn't be in `show_modified()` if they had the same OID, I believe.

File: dir.c [hunk 14]
```diff
@@ -1397,6 +1397,17 @@ enum pattern_match_result path_matches_pattern_list(
...
+	/*
+	 * Directory entries are matched if and only if a file
+	 * contained immediately within them is matched. For the
+	 * case of a directory entry, modify the path to create
+	 * a fake filename within this directory, allowing us to
+	 * use the file-base matching logic in an equivalent way.
+	 */
+	if (parent_pathname.len > 0 &&
+	    parent_pathname.buf[parent_pathname.len - 1] == '/')
+		strbuf_add(&parent_pathname, "-", 1);
```
Comment (dscho) [inline dir.c:1465]: Would `.` make sense here instead of `-`?

File: wt-status.c [hunk 15]
```diff
@@ -671,6 +699,28 @@ static void wt_status_collect_changes_initial(struct wt_status *s)
...
+			 * This is a sparse directory entry, so we want to collect all
+			 * of the added files within the tree. This requires recursively
+			 * expanding the trees to find the elements that are new in this
+			 * tree and marking them with DIFF_STATUS_ADDED.
+			 */
+			struct strbuf base = STRBUF_INIT;
+			struct pathspec ps;
+			struct tree *tree = lookup_tree(istate->repo, &ce->oid);
+
+			memset(&ps, 0, sizeof(ps));
```
Comment (dscho) [inline wt-status.c]: Maybe use the `= { 0 }` trick instead?

File: sparse-index.c [hunk 16]
```diff
@@ -186,6 +186,10 @@ int convert_to_sparse(struct index_state *istate)
 	cache_tree_free(&istate->cache_tree);
 	cache_tree_update(istate, 0);
 
+	istate->fsmonitor_has_run_once = 0;
+	FREE_AND_NULL(istate->fsmonitor_dirty);
+	FREE_AND_NULL(istate->fsmonitor_last_update);
+
```
Comment (dscho) [inline sparse-index.c:192]: An alternative would be to map the old extension data (which is essentially a compressed version of the `fsmonitor_dirty` bitmap, which we should be able to map relatively easily by iterating over the full and the sparse index simultaneously).

I do not suggest to implement this here, but maybe mention it as a potential future improvement?

File: wt-status.c [hunk 17]
```diff
@@ -664,13 +664,14 @@ static int add_file_to_list(const struct object_id *oid,
 	struct string_list_item *it;
 	struct wt_status_change_data *d;
 	struct wt_status *s = context;
-	char *full_name;
+	struct strbuf full_name = STRBUF_INIT;
```
Comment (dscho) [inline wt-status.c:667]: This is called potentially a _lot_ of times. If it turns out that we need to switch between sparse and full indices longer (or more often) than anticipated, we could pass in an initialized `strbuf` and only `strbuf_reset()` it in `add_file_to_list()`.

File: sparse-index.c [hunk 18]
```diff
@@ -261,7 +261,7 @@ void ensure_full_index(struct index_state *istate)
 			warning(_("index entry is a directory, but not sparse (%08x)"),
 				ce->ce_flags);
 
-		/* recursively walk into cd->name */
+		/* recursively walk into ce->name */
```
Comment (dscho) [inline sparse-index.c:264]: Since this typo got into upstream, do you maybe want to _not_ mark this as a `fixup!`?

**Review by dscho — APPROVED** [review]
Looks good!

I do have a fundamental question, though. If we can already say that a cache entry refers to a sparse dir by seeing whether its `ce_mode` says that it is a directory and its flags say `CE_SKIP_WORKTREE`, do we really need to append a `/` at the end? I ask because appending that slash seems to cause more trouble than it helps (sorting, binary searching, etc).

**Comment by derrickstolee** [comment]
> I do have a fundamental question, though. If we can already say that a cache entry refers to a sparse dir by seeing whether its `ce_mode` says that it is a directory and its flags say `CE_SKIP_WORKTREE`, do we really need to append a `/` at the end? I ask because appending that slash seems to cause more trouble than it helps (sorting, binary searching, etc).

While it makes things a bit more complicated in this scenario, you missed the part in the fundamentals where it is important for things like `index_name_pos()` when someone searches for a path within the sparse-directory entry. It is important that the sparse-directory entry fits in the correct position where its contained blobs would fit if the index was expanded.

File: sparse-index.c (hunk 18 above)
Comment (derrickstolee) [reply sparse-index.c:264]: That particular `fixup!` commit is more of a reminder to myself to include it in a cleanup patch somewhere down the line (not in this series)

File: dir.c (hunk 14 above)
Comment (derrickstolee) [reply dir.c:1465]: For some reason having a filename as `.` seems stranger to me than `-`. Perhaps `_` would seem less odd?

File: sparse-index.c (hunk 16 above)
Comment (derrickstolee) [reply sparse-index.c:192]: Are you recommending a translation of the old bitmap into a new bitmap based on the new path positions? Yes, that would work, but I don't think it is worth pursuing much at all. We are better off investing in the sparse-index work such that we only change the sparse directory entries during `git sparse-checkout set`, which is rare and likely is on a clean working directory.

File: wt-status.c (hunk 17 above)
Comment (derrickstolee) [reply wt-status.c:667]: The string is eventually added to a `struct string_list` that doesn't duplicate the strings. Thus, we will need to allocate these strings no matter what.

This case is also rare: we need to have the index differ from `HEAD` on a sparse directory entry, which requires strange situations. A performance drop in this situation is the least of the user's worries.

File: dir.c (hunk 14 above)
Comment (derrickstolee) [reply dir.c:1465]: I talked with @dscho offline here and here is the conclusion:

* The recommendation here is clever: `path/to/directory/.` is semantically identical to `path/to/directory`, so `.` is a natural option here.
* This relies on the cone-mode path-checking logic to be ignorant of this meaning of the trailing `/.`. While it _is_ dumb enough to ignore that, we would be adding an expectation that it remains dumb and otherwise doesn't check that the input path is a valid name for a file.

The conclusion is that this is a clever idea, but perhaps _too clever_.

Comment (jeffhostetler) [reply dir.c:1465]: I'm not sure there is a correct answer here.  But it makes me want to say that we want to treat the ce's as a {type,path} tuple with an accessor and all that and get away from the pattern of just iterating over a dumb list of strings (that all callers know is a dumb list of strings).

Is there existing code to detect file -vs- symlink collisions?  or file -vs- submodule collisions?  How are they handled?
I mean we're adding the concept of (sparse) directory nodes into the index; it would be nice if they collided the same way.  (I'm not trying to cause problems, just asking dumb questions.)

**Comment by derrickstolee** [comment]
I amended and force-pushed because some of the vfs-test jobs were stalled on unpacking the SDKs, but canceling the jobs did not give me a way to restart the jobs (the "Cancel Workflow" button was still up instead of "re-run all jobs")
